eddsa/eddsa.go

```diff
@@ -0,0 +1,163 @@
+package eddsa
+
+import (
+	"crypto/cipher"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/ed25519"
+	"github.com/dedis/crypto/random"
+)
+
+var suite = ed25519.NewAES128SHA256Ed25519(false)
+
+// EdDSA implements the EdDSA signature algorithm according to
+// the RFC https://tools.ietf.org/html/draft-josefsson-eddsa-ed25519-02
+type EdDSA struct {
+	seed   []byte
+	prefix []byte
+	// Secret being already hashed + bit tweaked
+	Secret abstract.Secret
+	// Public is the corresponding public key
+	Public abstract.Point
+}
+
+// NewEdDSAKey will return a freshly generated key pair to use for generating
+// EdDSA signatures.
+// If stream == nil, it will take the random.Stream.
+func NewEdDSA(stream cipher.Stream) *EdDSA {
+	if stream == nil {
+		stream = random.Stream
+	}
+	buffer := random.NonZeroBytes(32, stream)
+	scalar := sha512.Sum512(buffer)
+	scalar[0] &= 0xf8
+	scalar[31] &= 0x3f
+	scalar[31] |= 0x40
+
+	secret := suite.Secret().SetBytes(scalar[:32])
+	public := suite.Point().Mul(nil, secret)
+
+	return &EdDSA{buffer, scalar[32:], secret, public}
```

Comment:
as the variable-names are different from the field-names of `EdDSA`, I'd do the `Seed: scalar[32:]`-type initialisation of the structure.